iocore/nexthop/NextHopHost.h

```diff
@@ -0,0 +1,201 @@
+#include <atomic>
+#include "I_EventSystem.h"
+#include "PartitionedMap.h"
+#include "PropertyBlock.h"
+#include "ts/string_view.h"
+
+/**
+ * @file
+ *
+ * @brief this file defines structures to store data about each host and ip.
+ *
+ * HostAddr - stores a IpEndpoint, expire time, and up status. Additional properties can be defined at system start.
+ *
+ * HostName - stores a FQDN and a list of HostAddr. Additional properties can be defined at system start.
+ *  All references to HostNames will be shared pointers. They are only deleted when all references are deleted.
+ *
+ * Lookups are performed with one global mutex lock to protect integrity of the map.
+ * @see HostLookupByName, HostLookupByAddr
+ *
+ * When you are reading/writing a HostNamePtr, ensure that you find the assigned mutex from the LockPool. This ensures a single
+ * reader/writer to the HostName, associated HostAddrs and property blocks.
+ * @see HostName::getMutex()
+ *
+ */
+
+// Define a std::hash<>() for the lookmaps
+std_hasher_macro(IpEndpoint, ip, ats_ip_port_hash(ip));
+
+namespace NextHop
+{
+using Mutex = RWLock;
+
+/// uid for a host that should persist between updates and reloads.
+/// It is prefered for other systems to cache HostID instead of a HostNamePtr.
+using HostID = size_t;
```

Comment:
You can guarantee no collisions in this space?

Reply:
I was planning on hashing the URL to something like an uint64. : ( Yes, I suppose collisions are possible and terrible. There might still be a benefit to hashing internally. But I think your right, we can't rely on a hash to uniquely identify a host.

External systems should cache the host shared_ptr, or save the FQDN.